source/common/memory/debug.cc

```diff
@@ -0,0 +1,118 @@
+// Very simple memory debugging overrides for operator new/delete, to
+// help us quickly find simple memory violations:
+//   1. Double destruct
+//   2. Read before write (via scribbling)
+//   3. Read after delete (via scribbling)
+//
+// Note that valgrind does all of this much better, but is too slow to run all
+// the time. asan does read-after-delete detection but not read-before-init
+// detection. See
+// https://clang.llvm.org/docs/AddressSanitizer.html#initialization-order-checking
+// for more details.
+
+// Principle of operation: add 8 bytes to every allocation. The first
+// 4 bytes are a marker (LiveMarker1 or DeadMarker1). The next 4
+// bytes are used to store size of the allocation, which helps us
+// know how many bytes to scribble when we free.
+//
+// This code was adapted from mod_pagespeed, and adapted for Envoy
+// style. Original source:
+// https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/mem_debug.cc
+
+// We keep a global count of bytes allocated so that memory-consumption tests
+// work with memory debugging. Put another way, if we disable tcmalloc when
+// compiling for debug, we want the memory-debugging tests to work, otherwise we
+// can't debug them.
+#include "common/memory/debug.h"
+
+#include <atomic>
+#include <cassert> // don't use Envoy ASSERT as it may allocate memory.
+#include <cstdlib>
+
+#include "common/memory/align.h"
+
+static std::atomic<int64_t> bytes_allocated(0);
+
+namespace Envoy {
+namespace Memory {
+
+// We always provide the constructor entry-point to be called to force-load this
+// module, regardless of compilation mode. If the #ifdefs line as required
+// below, then it will also override operator new/delete in various flavors so
+// that we can debug memory issues.
+Debug::Debug() = default;
+
+// We also provide the bytes-loaded counter, though this will return 0 when
+// memory-debugging is not compiled in.
+uint64_t Debug::bytesUsed() { return uint64_t(bytes_allocated); }
+
+} // namespace Memory
+} // namespace Envoy
+
+#ifdef MEMORY_DEBUG_ENABLED
+
+namespace {
+
+constexpr uint32_t LiveMarker1 = 0xfeedface;         // first 4 bytes after alloc
+constexpr uint64_t LiveMarker2 = 0xfeedfacefeedface; // first 4 bytes after alloc
+constexpr uint32_t DeadMarker1 = 0xabacabff;         // first 4 bytes after free
```

Comment:
I cannot comment on the original thread, so reviving it here.

> What's the value of having 2 different dead markers? Is that useful in practice?

> It has been occasionally useful, yes, to to distinguish between blobs of memories that are allocated but initialized vs freed.
>
> note also that tcmalloc's debugallocation does the same thing (with different patterns).

Yes, the usefulness of `Live` and `Dead` is quite obvious... but my question was about having different `Marker1` and `Marker2` (i.e. marking first 4 bytes of the freshly allocated / freed memory differently from the rest).